packages/telemetry/src/context-aware-slog.js

```diff
@@ -0,0 +1,475 @@
+     */
+    persistContext: context => {
+      /** @type {Array<string>} */
+      const keys = [];
+      const values = [];
+
+      for (const [key, value] of Object.entries(context)) {
+        keys.push(`[${key}]`);
+        if (value === undefined) values.push(null);
+        else
+          values.push(
+            DATABASE_SCHEMA[key].startsWith(DATABASE_TYPES.TEXT)
+              ? `'${value}'`
+              : value,
+          );
+      }
+
+      databaseInstance
+        .prepare(
+          `INSERT OR REPLACE INTO '${DATABASE_TABLE_NAME}' (key, ${keys.join(', ')})
+           VALUES ('${DATABASE_TRIGGER_CONTEXT_KEY_NAME}', ${values.join(', ')})`,
+        )
+        .run();
+    },
+
+    restoreContext: () => {
+      /**
+       * @type {Context & {key?: string} | undefined}
+       */
+      // @ts-expect-error
+      const row = databaseInstance
+        .prepare(
+          `SELECT * FROM '${DATABASE_TABLE_NAME}' WHERE key = '${DATABASE_TRIGGER_CONTEXT_KEY_NAME}'`,
+        )
+        .get();
+
+      delete row?.key;
+
+      return row || null;
+    },
+  };
+};
+
+const stringify = data =>
+  JSON.stringify(data, (_, value) =>
+    typeof value === 'bigint' ? Number(value) : value,
+  );
+
+/**
+ *
+ * @param {{env: typeof process.env}} options
+ */
+export const makeSlogSender = async options => {
+  const { CHAIN_ID, OTEL_EXPORTER_OTLP_ENDPOINT } = options.env;
+  if (!OTEL_EXPORTER_OTLP_ENDPOINT)
+    return console.warn(
+      'Ignoring invocation of slogger "context-aware-slog" without the presence of "OTEL_EXPORTER_OTLP_ENDPOINT" envrionment variable',
+    );
+
+  const loggerProvider = new LoggerProvider({
+    resource: new Resource(getResourceAttributes(options)),
+  });
+  loggerProvider.addLogRecordProcessor(
+    new SimpleLogRecordProcessor(new OTLPLogExporter({ keepAlive: true })),
+  );
+
+  logs.setGlobalLoggerProvider(loggerProvider);
+  const logger = logs.getLogger('default');
+
+  const db = await getDatabaseInstance();
+
+  /** @type Array<Context | null> */
+  let [
+    blockContext,
+    crankContext,
+    initContext,
+    lastPersistedTriggerContext,
+    replayContext,
+    triggerContext,
+  ] = [null, null, null, null, null, null];
+
+  /**
+   * @param {Context} context
+   */
+  const persistContext = context => {
+    lastPersistedTriggerContext = context;
+    db.persistContext(context);
+  };
+
+  const restoreContext = () => {
+    if (lastPersistedTriggerContext) return lastPersistedTriggerContext;
+    return db.restoreContext();
+  };
+
+  /**
+   * @param {Slog} slog
+   */
+  const slogSender = ({ monotime, time: timestamp, ...body }) => {
+    const finalBody = { ...body };
+
+    /** @type {LogAttributes} */
+    let logAttributes = {
+      'chain-id': String(CHAIN_ID),
+      'process.uptime': monotime,
+      timestamp,
+    };
+
+    /**
+     * Add any before report operations here
+     * like setting context data
+     */
+    switch (body.type) {
```

Comment:
As a matter of code organization, I would prefer to see the block and context building types handled first so that reading we can more clearly see each being built up as they would when processing the slog events. Here is seems they're somewhat lexicographically sorted, which doesn't really bring any value.

Reply:
Can you suggest a rough order? I am not sure which cases should be handled in what order for better understanding

Reply:
The order in the issue is roughly what I would expect. I'll double check that it fits

Reply:
I checked and the order in the issue is what I would expect to see here.